backend/src/models/shared.rs

```diff
@@ -0,0 +1,135 @@
+use super::prelude::*;
+use std::result::Result;
+
+#[derive(ToSchema, Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct Hostname(
+    #[serde(
+        deserialize_with = "hyper_serde::deserialize",
+        serialize_with = "hyper_serde::serialize"
+    )]
+    Uri,
+);
+
+#[derive(Debug, Error)]
+pub enum HostnameError {
+    #[error("bad address: no port")]
+    NoPort,
+    #[error("bad address: couldn't parse hostname")]
+    BadAddress,
+}
+
+impl Hostname {
+    #[must_use]
+    pub fn port(&self) -> Option<u16> {
+        self.0.port_u16()
+    }
+
+    /// Creates [`Hostname`] from string with specified port
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if address doesn't have a port or a port is invalid
+    pub fn with_port(address: &str, port: u16) -> error_stack::Result<Self, HostnameError> {
+        let (body, _) = address.rsplit_once(':').ok_or(HostnameError::NoPort)?;
+        let mut body = body.to_string();
+        body.push_str(&format!(":{port}"));
+
+        Ok(Self(
+            hyper::http::Uri::from_str(&body).change_context(HostnameError::BadAddress)?,
+        ))
+    }
+}
+
+impl TryFrom<SocketAddr> for Hostname {
+    type Error = <Uri as TryFrom<String>>::Error;
+
+    fn try_from(value: SocketAddr) -> Result<Self, Self::Error> {
+        Ok(Self(Uri::try_from(value.to_string())?))
+    }
+}
+
+impl TryFrom<Hostname> for SocketAddr {
+    type Error = std::net::AddrParseError;
+
+    fn try_from(value: Hostname) -> Result<Self, Self::Error> {
+        value.to_string().parse()
+    }
+}
+
+impl ToString for Hostname {
+    fn to_string(&self) -> String {
+        self.0.to_string()
+    }
+}
+
+/// Data needed to connect to a BOB cluster
+#[derive(IntoParams, ToSchema, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[schema(example = json!({"hostname": "0.0.0.0:7000", "credentials": {"login": "archeoss", "password": "12345"}}))]
+pub struct BobConnectionData {
+    /// Address to connect to
+    pub hostname: Hostname,
+
+    /// [Optional] Credentials used for BOB authentication
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub credentials: Option<Credentials>,
+}
+
+/// Optional auth credentials for a BOB cluster
+#[derive(
+    ToSchema, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Default, Serialize, Deserialize,
```

Comment:
Please, write custom implementation for `Debug` so that it does not write `password` to the output (this password can easily leak somewhere)